tmt/steps/prepare/artifact/providers/repository_provider.py

```diff
@@ -1,108 +1,144 @@
-        """The location (URL) of the repository file"""
-        return self._raw_artifact
+    def filename(self) -> str:
+        """A suitable filename extracted from the URL path."""
+        path = urlparse(self.url).path
+        return unquote(Path(path).name)
 
 
-class RepositoryFileProvider(ArtifactProvider[RepositoryFileInfo]):
+class RepositoryFileProvider(ArtifactProvider[RpmArtifactInfo]):
     """
-    Provides repository files for download.
+    Sets up a repository on the guest and lists the RPMs it provides.
+
+    The artifact_id is a URL to a .repo file. This provider's main
+    purpose is to download this file and place it in '/etc/yum.repos.d/'.
 
-    The artifact_id is expected to be a URL to a repository file, which will be
-    downloaded to a specified destination on a guest system. The provider
-    ensures the URL points to a valid ``.repo`` file and handles downloading
-    it to the guest.
+    It then lists all RPMs made available by the new repository but does
+    not download any of them.
     """
 
-    # TODO: Change to RepositoryFileProvider(ArtifactProvider[RpmArtifactInfo])
-    # because we will be listing RPM's
+    def __init__(self, logger: tmt.log.Logger, artifact_id: str):
+        super().__init__(logger, artifact_id)
+        self.repo_file = RepositoryFile(url=self.artifact_id)
+        # Cache for the list of RPMs discovered in the repository
+        self._rpm_list: Optional[list[dict[str, Any]]] = None
+        self._repository_installed = False
 
     def _parse_artifact_id(self, artifact_id: str) -> str:
-        # Validate the artifact_id by creating a RepositoryFileInfo instance
-        RepositoryFileInfo(_raw_artifact=artifact_id)
+        """
+        Validate the artifact identifier using the RepositoryFile class.
+        """
+        # The constructor of RepositoryFile handles URL validation
+        RepositoryFile(url=artifact_id)
         return artifact_id
 
-    def list_artifacts(self) -> Iterator[RepositoryFileInfo]:
-        yield RepositoryFileInfo(_raw_artifact=self.artifact_id)
+    def _fetch_rpms(self, guest: Guest, repo_filepath: Path) -> None:
+        """
+        Query the guest to find all packages available in the new repository.
+        """
+        # TODO: This method needs to be implemented to populate self._rpm_list with RPMs
+        # from the repository. Currently using an empty list as a temporary solution.
+        self._rpm_list = []
+
+    def list_artifacts(self) -> Iterator[RpmArtifactInfo]:
+        """
+        List all RPMs available from the repository.
+
+        Note: This requires the repository to be installed and queried first,
+        which is handled by the `download_artifacts` method.
+        """
+        if self._rpm_list is None:
+            raise GeneralError(
+                "RPM list not fetched. Call 'download_artifacts' to install "
+                "the repository and populate the package list."
+            )
```

Comment:
If we would ever hit this, it is because we have messed up the implementation
```suggestion
        assert self._rpm_list is not None
```
But I would also make the check against `_repository_installed` so that `self._rpm_list` can always be a `list`